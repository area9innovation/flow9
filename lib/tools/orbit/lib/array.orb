// Array utility functions library for Orbit

// map: Apply a function to each element in an array
fn map(arr, f) = (
    if length(arr) = 0 then []
    else [f(arr[0])] + map(subrange(arr, 1, length(arr) - 1), f)
);

// filter: Keep only elements that match a predicate
fn filter(arr, pred) = (
    if length(arr) = 0 then []
    else if pred(arr[0]) then 
        [arr[0]] + filter(subrange(arr, 1, length(arr) - 1), pred)
    else 
        filter(subrange(arr, 1, length(arr) - 1), pred)
);

// fold: Accumulate a result based on each array element and an accumulator
fn fold(arr, init, f) = (
    if length(arr) = 0 then init
    else fold(subrange(arr, 1, length(arr) - 1), f(init, arr[0]), f)
);

// mapi: Map with index
fn mapi(arr, f) = (
    fn mapiHelper(arr, f, idx) = (
        if length(arr) = 0 then []
        else [f(arr[0], idx)] + mapiHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1)
    );
    mapiHelper(arr, f, 0)
);

// filteri: Filter with index
fn filteri(arr, pred) = (
    fn filteriHelper(arr, pred, idx) = (
        if length(arr) = 0 then []
        else if pred(arr[0], idx) then 
            [arr[0]] + filteriHelper(subrange(arr, 1, length(arr) - 1), pred, idx + 1)
        else 
            filteriHelper(subrange(arr, 1, length(arr) - 1), pred, idx + 1)
    );
    filteriHelper(arr, pred, 0)
);

// foldi: Fold with index
fn foldi(arr, init, f) = (
    fn foldiHelper(arr, acc, f, idx) = (
        if length(arr) = 0 then acc
        else foldiHelper(subrange(arr, 1, length(arr) - 1), f(acc, arr[0], idx), f, idx + 1)
    );
    foldiHelper(arr, init, f, 0)
);

// tail: Return all elements except the first one
fn tail(arr) = (
    if length(arr) <= 1 then []
    else subrange(arr, 1, length(arr) - 1)
);

// tailFrom: Return all elements from a given index to the end
fn tailFrom(arr, idx) = (
    if idx >= length(arr) then []
    else subrange(arr, idx, length(arr) - idx)
);

// take: Return the first n elements of an array
fn take(arr, n) = (
    if n <= 0 || length(arr) = 0 then []
    else [arr[0]] + take(subrange(arr, 1, length(arr) - 1), n - 1)
);

// iter: Call a function for each element (for side effects)
fn iter(arr, f) = (
    if length(arr) = 0 then 0 // Return 0 instead of ()
    else (
        f(arr[0]);
        iter(subrange(arr, 1, length(arr) - 1), f)
    )
);

// iteri: Iterate with index
fn iteri(arr, f) = (
    fn iteriHelper(arr, f, idx) = (
        if length(arr) = 0 then 0 // Return 0 instead of ()
        else (
            f(arr[0], idx);
            iteriHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1)
        )
    );
    iteriHelper(arr, f, 0)
);

// filtermap: Combine filter and map operations
fn filtermap(arr, f) = (
    if length(arr) = 0 then []
    else (
        let result = f(arr[0]);
        result is (
            Some(val) => [val] + filtermap(subrange(arr, 1, length(arr) - 1), f);
            None() => filtermap(subrange(arr, 1, length(arr) - 1), f);
            _ => filtermap(subrange(arr, 1, length(arr) - 1), f) // For safety
        )
    )
);

// filtermapi: Filtermap with index
fn filtermapi(arr, f) = (
    fn filtermapHelper(arr, f, idx) = (
        if length(arr) = 0 then []
        else (
            let result = f(arr[0], idx);
            result is (
                Some(val) => [val] + filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1);
                None() => filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1);
                _ => filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1) // For safety
            )
        )
    );
    filtermapHelper(arr, f, 0)
);

// removeIndex: Remove an element at a specific index
fn removeIndex(arr, idx) = (
    if idx < 0 || idx >= length(arr) then arr
    else take(arr, idx) + tailFrom(arr, idx + 1)
);

// removeRange: Remove a range of elements
fn removeRange(arr, start, count) = (
    if start < 0 || start >= length(arr) || count <= 0 then arr
    else take(arr, start) + tailFrom(arr, start + count)
);

// removeFirst: Remove the first occurrence of a specific element
fn removeFirst(arr, element) = (
    fn removeFirstHelper(before, after) = (
        if length(after) = 0 then before
        else if after[0] = element then before + subrange(after, 1, length(after) - 1)
        else removeFirstHelper(before + [after[0]], subrange(after, 1, length(after) - 1))
    );
    removeFirstHelper([], arr)
);

// removeAll: Remove all occurrences of a specific element
fn removeAll(arr, element) = (
    filter(arr, \x.x != element != 0)
);

// contains: Check if an element exists in the array
fn contains(arr, element) = (
    if length(arr) = 0 then false
    else if arr[0] = element then true
    else contains(subrange(arr, 1, length(arr) - 1), element)
);

// exists: Check if any element satisfies a predicate
fn exists(arr, pred) = (
    if length(arr) = 0 then false
    else if pred(arr[0]) then true
    else exists(subrange(arr, 1, length(arr) - 1), pred)
);

// forall: Check if all elements satisfy a predicate
fn forall(arr, pred) = (
    if length(arr) = 0 then true
    else if !pred(arr[0]) then false
    else forall(subrange(arr, 1, length(arr) - 1), pred)
);

// uniq: Remove duplicates from an array
fn uniq(arr) = (
    if length(arr) <= 1 then arr
    else (
        fn uniq_helper(remaining, result) = (
            if length(remaining) = 0 then result
            else (
                let current = remaining[0];
                let rest = tail(remaining);
                
                if contains(result, current) then
                    // Skip duplicate
                    uniq_helper(rest, result)
                else
                    // Keep this element
                    uniq_helper(rest, result + [current])
            )
        );
        uniq_helper(arr, [])
    )
);