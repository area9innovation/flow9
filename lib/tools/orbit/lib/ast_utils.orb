// AST utility functions for Orbit

// Create a properly nested AST node by folding over an array of arguments
// This function handles arrays of any length, whereas makeAst expects exactly 2 arguments
fn makeFoldedAst(op, args) = (
    let len = length(args);
    
    if len = 0 then (
        // Handle empty array case - return an appropriate default value
        // This depends on the operator, but we'll return a neutral element
        if op = "&&" then true
        else if op = "||" then false
        else args  // Return empty array as fallback
    ) else if len = 1 then (
        // For single element, just return it without creating an operation
        args[0]
    ) else if len = 2 then (
        // For exactly two elements, we can use makeAst directly
        makeAst(op, args)
    ) else (
        // For 3+ elements, fold to create a proper binary tree
        // Start with the first element and fold over the rest
        let initialAst = args[0];
        let remainingArgs = subrange(args, 1, len - 1);
        
        // Fold over the remaining arguments, combining each with the accumulator
        fold(remainingArgs, initialAst, \acc, element -> (
            makeAst(op, [acc, element])
        ))
    )
);