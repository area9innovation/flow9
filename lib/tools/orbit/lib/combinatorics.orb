// Basic combinatorial functions
fn factorial(n) = (
	if n <= 1 then 1
	else n * factorial(n - 1)
);

fn binomial_coefficient(n, k) = (
	if k < 0 || k > n then 0
	else if k == 0 || k == n then 1
	else (
		// Use multiplicative formula to avoid large intermediate values
		fn binom_helper(i, result) = (
			if i > k then result
			else binom_helper(i + 1, result * (n - k + i) / i)
		);
		binom_helper(1, 1)
	)
);

fn permutations(n, r) = (
	if r < 0 || r > n then 0
	else if r == 0 then 1
	else (
		// Calculate n! / (n-r)!
		fold(range(n-r+1, n+1), 1, \product, i.product * i)
	)
);

// Generate permutations of an array
fn generate_permutations(arr) = (
	if length(arr) <= 1 then [arr]
	else (
		let result = [];
		fn perms_helper(i, perms) = (
			if i >= length(arr) then perms
			else (
				let element = arr[i];
				let rest = remove_at_index(arr, i);
				let sub_perms = generate_permutations(rest);

				let new_perms = fold(sub_perms, perms, \acc, p.acc + [[element] + p]);
				perms_helper(i + 1, new_perms)
			)
		);
		perms_helper(0, [])
	)
);

// Generate combinations of an array
fn generate_combinations(arr, r) = (
	if r > length(arr) then []
	else if r == 0 then [[]]
	else if r == length(arr) then [arr]
	else (
		if length(arr) == 0 then []
		else (
			let first = arr[0];
			let rest = subrange(arr, 1, length(arr) - 1);

			// Combinations that include the first element
			let with_first = map(generate_combinations(rest, r - 1),
													\comb.[first] + comb);

			// Combinations that exclude the first element
			let without_first = generate_combinations(rest, r);

			with_first + without_first
		)
	)
);
