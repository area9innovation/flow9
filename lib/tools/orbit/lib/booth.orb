import lib/array;

// Linear time Booth's algorithm to find canonical form in cyclic groups
fn booth_canonical(arr) = (
	if length(arr) <= 1 then arr
	else (
		let n = length(arr);

		// Create an auxiliary array which is the original array concatenated with itself
		let double_arr = arr + arr;

		// Initialize failure function array with -1
		// (We use -1 instead of conventional 0 to simplify logic)
		let f = \i. if i < 0 || i >= n then -1 else f(i);

		// Compute the min rotation index in linear time
		fn compute_min_index() = (
			// Initialize starting values
			fn find_min(i, j, k) = (
				// We're done when we've checked all rotations
				if i + k >= 2 * n then j
				else if k >= n then find_min(i + 1, j, 0)
				else (
					// Compute the current indices we're comparing
					let i_idx = (i + k) % (2 * n);
					let j_idx = (j + k) % (2 * n);

					if double_arr[i_idx] < double_arr[j_idx] then
						// i rotation is smaller, j becomes i
						find_min(i, i, k + 1)
					else if double_arr[i_idx] > double_arr[j_idx] then
						// j rotation is smaller, i moves forward
						find_min(i + k + 1, j, 0)
					else
						// Equal so far, continue comparison
						find_min(i, j, k + 1)
				)
			);

			// Start with first two positions
			find_min(0, 0, 0)
		);

		let min_start = compute_min_index();

		// Build the canonical form starting from min_start
		fn build_result(idx, result) = (
			if idx >= n then result
			else build_result(idx + 1, result + [arr[(min_start + idx) % n]])
		);

		build_result(0, [])
	)
);

