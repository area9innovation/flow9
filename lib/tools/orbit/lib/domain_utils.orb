import lib/graph;

// Compare two domains in the hierarchy
// Returns: 
//   -1 if a is more specific than b (a ⊂ ... ⊂ b)
//    0 if a and b are the same
//    1 if b is more specific than a (b ⊂ ... ⊂ a)
//    None() if a and b are not directly related
fn compareDomains(graph, a, b) = (
    if (a == b) then 0  // Same domain
    else if (isReachable(graph, a, b)) then -1  // a is more specific than b
    else if (isReachable(graph, b, a)) then 1   // b is more specific than a
    else None()  // Not directly related
);

// Find the lowest domain from a list of domains
// The lowest domain is the most specific one in the hierarchy
fn findLowestDomain(graph, domains) = (
    if (length(domains) == 0) then None()
    else if (length(domains) == 1) then Some(domains[0])
    else (
        // Start with the first domain as our current lowest
        let initialLowest = domains[0];
        
        // Compare with each other domain to find the lowest
        fold(subrange(domains, 1, length(domains) - 1), initialLowest, \currentLowest, domain -> (
            compareDomains(graph, domain, currentLowest) is (
                -1 => domain;  // This domain is more specific than current lowest
                0 => currentLowest;  // Same domain, keep current
                1 => currentLowest;  // Current lowest is more specific
                None() => (
                    // Domains are not directly related, we keep the current lowest
                    // A more sophisticated approach might look at the full hierarchy
                    currentLowest
                )
            )
        ))
    )
);

// Get the most specific domain a node belongs to
// This is the main function to be used in the OGraph system
fn getNodeMostSpecificDomain(domainGraph, og, nodeId) = (
    // Get all domains the node belongs to
    let domains = getONodeBelongsTo(og, nodeId);
    
    // Find and return the lowest domain from the list
    findLowestDomain(domainGraph, domains)
);

"Domain Hierarchy Utility Library Successfully Loaded"