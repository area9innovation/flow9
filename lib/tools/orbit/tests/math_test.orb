// Test suite demonstrating simplification and canonicalization of various mathematical expressions
// using the Orbit OGraph system. Assumes relevant rewrite rules from math libraries are loaded.

import lib/array;
import lib/rewrite;
import lib/complex; // For 'i' and Complex constructor if needed directly
// Other imports might be needed depending on how rules are structured/loaded

// --- Helper Function to Simplify and Print ---
fn simplify_and_print(label, expr) = (
    println("--- " + label + " ---");
    println("Original:     " + prettyOrbit(expr));
    // Apply simplification rules (assuming they are loaded into all_math_rules)
    // Increase iterations for complex cases
    let simplified = applyRulesUntilFixedPoint(expr, all_math_rules, 50);
    println("Simplified:   " + prettyOrbit(simplified));
    println("");
);

// --- List of Test Expressions ---

let test_expressions = [
    // --- Basic Algebra ---
    Pair("Basic Arithmetic 1", quote(5 + 3 * 2 - 1)),
    Pair("Basic Arithmetic 2", quote(0 + x + 1 * y - 0)),
    Pair("Associativity Add", quote(a + (b + c))), // Should canonicalize to (a+b)+c or (+ a b c)
    Pair("Associativity Mul", quote((a * b) * c)), // Should canonicalize to a*(b*c) or (* a b c)
    Pair("Commutativity Add", quote(b + a)),       // Should canonicalize to a+b (if a<b)
    Pair("Commutativity Mul", quote(y * x)),       // Should canonicalize to x*y (if x<y)
    Pair("Distributivity 1", quote(a * (b + c))),
    Pair("Distributivity 2", quote((x + y) * z)),
    Pair("Subtraction Def", quote(a - b)),
    Pair("Double Negative", quote(-(-a))),
    Pair("Negative Multiplication", quote((-a) * (-b))),
    Pair("Zero Product", quote(x * 0)),

    // --- Polynomials (Requires GLEX rules from polynomials.orb) ---
    Pair("Polynomial Collect 1", quote(3*x*y + 2*y*x + 5)), // Expect 5*x*y + 5
    Pair("Polynomial Collect 2", quote(x^2 + 3*x + 2*x^2 - x)), // Expect 3*x^2 + 2*x
    Pair("Polynomial GLEX Order 1", quote(z + y^2 + x^3)), // Expect x^3 + y^2 + z
    Pair("Polynomial GLEX Order 2", quote(c*a + b*a + a^2)), // Expect a^2 + a*b + a*c (factor sort)
    Pair("Polynomial Distribute", quote(x * (y + z^2))),

    // --- Calculus (Symbolic Differentiation - requires diff rules) ---
    Pair("Simple Derivative", quote(diff(x^2 + 3*x + 5, x))), // Expect 2*x + 3
    Pair("Product Rule Derivative", quote(diff(x * sin(x), x))), // Expect 1*sin(x) + x*cos(x)
    Pair("Chain Rule Derivative", quote(diff(cos(x^3), x))), // Expect -sin(x^3) * 3*x^2
    Pair("Constant Derivative", quote(diff(10, x))), // Expect 0

    // --- Calculus (Symbolic Integration - requires integration rules, might be complex) ---
    Pair("Simple Integral", quote(integrate(2*x + 3, x))), // Expect x^2 + 3*x + C (Constant C handling needed)
    Pair("Power Rule Integral", quote(integrate(x^3, x))), // Expect x^4 / 4 + C

    // --- Complex Numbers (Requires rules from complex.orb and math.orb) ---
    Pair("Complex Addition", quote(Complex(1.0, 2.0) + Complex(3.0, -1.0))), // Expect Complex(4.0, 1.0)
    Pair("Complex Multiplication", quote(Complex(2.0, 1.0) * Complex(1.0, 3.0))), // Expect Complex(-1.0, 7.0)
    Pair("Imaginary Unit Squared", quote(i * i)), // Expect Complex(-1.0, 0.0) or -1.0
    Pair("Complex Conjugate", quote(conjugateC(Complex(4.0, 5.0)))), // Expect Complex(4.0, -5.0)
    Pair("Euler's Formula", quote(expC(i * pi))), // Expect Complex(-1.0, 0.0) or -1.0
    Pair("Complex Real Check", quote(Complex(5.0, 0.0))), // Expect 5.0
    Pair("Magnitude Squared", quote(magnitudeC(Complex(3.0, 4.0))^2)), // Expect 25.0 (via z*conj(z) rule?)

    // --- Linear Algebra (Matrices/Vectors - requires linalg.orb rules) ---
    Pair("Matrix Transpose", quote(transposeM(transposeM(A)))), // Expect A
    Pair("Matrix Distributivity", quote(A * (B + C))), // Expect A*B + A*C
    Pair("Transpose Product", quote(transposeM(A * B))), // Expect B^T * A^T (needs explicit transpose func)
    Pair("Matrix Identity Mul", quote(matrix_multiply(A, identityMatrix(n)))), // Needs identityMatrix function

    // --- Number Theory (Requires number_theory.orb rules) ---
    Pair("Modular Exponentiation", quote(modPow(3, 5, 7))), // 3^5 mod 7 = 243 mod 7 = 5
    Pair("Fermat's Little Theorem", quote(modPow(2, 11, 11))), // Expect 2 % 11 = 2 (if rule loaded)
    Pair("GCD Euclidean", quote(gcd(48, 18))), // Expect 6

    // --- Trigonometry (Requires rules from math.orb) ---
    Pair("Pythagorean Identity", quote(cos(theta)^2 + sin(theta)^2)), // Expect 1.0
    Pair("Double Angle Sin", quote(sin(2.0 * alpha))), // Expect 2.0*sin(alpha)*cos(alpha)
    Pair("Angle Sum Cos", quote(cos(x + y))), // Expect cos(x)*cos(y) - sin(x)*sin(y)
    Pair("Cos Neg Arg", quote(cos(-z))), // Expect cos(z)
    Pair("Exp/Trig Connection", quote(expC(i * x) + expC(-i * x))), // Expect 2.0 * cos(x) (via complex arithmetic)

    // --- Limits (Requires limit rules from math.orb) ---
    Pair("Limit Simple", quote(limit(1.0 / k, k, infinity))), // Expect 0.0
    Pair("Limit Sum", quote(limit(5 + 1.0/k, k, infinity))), // Expect 5.0

    // --- Mixed Expressions ---
    Pair("Mixed Trig/Poly", quote(diff(sin(x^2), x))), // Expect cos(x^2)*2*x
    Pair("Mixed Complex/Exp", quote(logC(expC(Complex(1.0, pi/4.0))))) // Expect Complex(1.0, pi/4.0)
];

// --- Run the Tests ---
println("Comprehensive Math Simplification Tests");
println("=====================================\\n");

iter(test_expressions, \pair -> (
    pair is (Pair(label, expr) => simplify_and_print(label, expr))
));

println("\\n--- Tests Complete ---");
