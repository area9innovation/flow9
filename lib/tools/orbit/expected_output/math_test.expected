/home/alstrup/area9/flow9/lib/tools/orbit/lib/complex.orb: /home/alstrup/area9/flow9/lib/tools/orbit/lib/complex.orb:211: Parse error
Comprehensive Math Simplification Tests
=====================================\n
--- Basic Arithmetic 1 ---
Original:     ((5 + (3 * 2)) - 1)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   ((5 + (3 * 2)) - 1)
--- Basic Arithmetic 2 ---
Original:     (((0 + x) + (1 * y)) - 0)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (((0 + x) + (1 * y)) - 0)
--- Associativity Add ---
Original:     (a + (b + c))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (a + (b + c))
--- Associativity Mul ---
Original:     ((a * b) * c)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   ((a * b) * c)
--- Commutativity Add ---
Original:     (b + a)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (b + a)
--- Commutativity Mul ---
Original:     (y * x)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (y * x)
--- Distributivity 1 ---
Original:     (a * (b + c))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (a * (b + c))
--- Distributivity 2 ---
Original:     ((x + y) * z)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   ((x + y) * z)
--- Subtraction Def ---
Original:     (a - b)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (a - b)
--- Double Negative ---
Original:     --a
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   --a
--- Negative Multiplication ---
Original:     (-a * -b)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (-a * -b)
--- Zero Product ---
Original:     (x * 0)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (x * 0)
--- Polynomial Collect 1 ---
Original:     ((((3 * x) * y) + ((2 * y) * x)) + 5)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   ((((3 * x) * y) + ((2 * y) * x)) + 5)
--- Polynomial Collect 2 ---
Original:     ((((x^2) + (3 * x)) + (2 * (x^2))) - x)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   ((((x^2) + (3 * x)) + (2 * (x^2))) - x)
--- Polynomial GLEX Order 1 ---
Original:     ((z + (y^2)) + (x^3))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   ((z + (y^2)) + (x^3))
--- Polynomial GLEX Order 2 ---
Original:     (((c * a) + (b * a)) + (a^2))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (((c * a) + (b * a)) + (a^2))
--- Polynomial Distribute ---
Original:     (x * (y + (z^2)))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (x * (y + (z^2)))
--- Simple Derivative ---
Original:     diff((((x^2) + (3 * x)) + 5), x)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   diff((((x^2) + (3 * x)) + 5), x)
--- Product Rule Derivative ---
Original:     diff((x * sin(x)), x)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   diff((x * sin(x)), x)
--- Chain Rule Derivative ---
Original:     diff(cos((x^3)), x)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   diff(cos((x^3)), x)
--- Constant Derivative ---
Original:     diff(10, x)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   diff(10, x)
--- Simple Integral ---
Original:     integrate(((2 * x) + 3), x)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   integrate(((2 * x) + 3), x)
--- Power Rule Integral ---
Original:     integrate((x^3), x)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   integrate((x^3), x)
--- Complex Addition ---
Original:     (Complex(1, 2) + Complex(3, -1))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (Complex(1, 2) + Complex(3, -1))
--- Complex Multiplication ---
Original:     (Complex(2, 1) * Complex(1, 3))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (Complex(2, 1) * Complex(1, 3))
--- Imaginary Unit Squared ---
Original:     (i * i)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (i * i)
--- Complex Conjugate ---
Original:     conjugateC(Complex(4, 5))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   conjugateC(Complex(4, 5))
--- Euler's Formula ---
Original:     expC((i * pi))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   expC((i * pi))
--- Complex Real Check ---
Original:     Complex(5, 0)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   Complex(5, 0)
--- Magnitude Squared ---
Original:     (magnitudeC(Complex(3, 4))^2)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (magnitudeC(Complex(3, 4))^2)
--- Matrix Transpose ---
Original:     transposeM(transposeM(A))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   transposeM(transposeM(A))
--- Matrix Distributivity ---
Original:     (A * (B + C))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (A * (B + C))
--- Transpose Product ---
Original:     transposeM((A * B))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   transposeM((A * B))
--- Matrix Identity Mul ---
Original:     matrix_multiply(A, identityMatrix(n))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   matrix_multiply(A, identityMatrix(n))
--- Modular Exponentiation ---
Original:     modPow(3, 5, 7)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   modPow(3, 5, 7)
--- Fermat's Little Theorem ---
Original:     modPow(2, 11, 11)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   modPow(2, 11, 11)
--- GCD Euclidean ---
Original:     gcd(48, 18)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   gcd(48, 18)
--- Pythagorean Identity ---
Original:     ((cos(theta)^2) + (sin(theta)^2))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   ((cos(theta)^2) + (sin(theta)^2))
--- Double Angle Sin ---
Original:     sin((2 * alpha))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   sin((2 * alpha))
--- Angle Sum Cos ---
Original:     cos((x + y))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   cos((x + y))
--- Cos Neg Arg ---
Original:     cos(-z)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   cos(-z)
--- Exp/Trig Connection ---
Original:     (expC((i * x)) + expC((-i * x)))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   (expC((i * x)) + expC((-i * x)))
--- Limit Simple ---
Original:     limit((1 / k), k, infinity)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   limit((1 / k), k, infinity)
--- Limit Sum ---
Original:     limit((5 + (1 / k)), k, infinity)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   limit((5 + (1 / k)), k, infinity)
--- Mixed Trig/Poly ---
Original:     diff(sin((x^2)), x)
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   diff(sin((x^2)), x)
--- Mixed Complex/Exp ---
Original:     logC(expC(Complex(1, (pi / 4))))
Variable not found: all_math_rules
ERROR: Expected array in arrayLength, got OrInt
Simplified:   logC(expC(Complex(1, (pi / 4))))
\n--- Tests Complete ---
Result: 0

VERIFY - THIS EXPECTED OUTPUT NEEDS HUMAN VERIFICATION (Generated Sun May  4 14:32:45 CEST 2025)
